packages/react-paypal-js/src/v6/components/PayPalProvider.tsx

```diff
@@ -0,0 +1,247 @@
+ * - Client: Initializes in PENDING state, loads scripts on mount
+ * - Hydration: Client takes over from server state seamlessly
+ *
+ * Props are automatically deep-compared to prevent unnecessary SDK reloads.
+ * You can pass options directly without manual memoization.
+ *
+ * @example
+ * <PayPalSdkInstanceProvider
+ *   components={["paypal-payments"]}
+ *   clientToken={token}
+ *   scriptOptions={{ environment: "sandbox" }}
+ * >
+ *   {children}
+ * </PayPalSdkInstanceProvider>
+ */
+export const PayPalProvider: React.FC<PayPalProviderProps> = ({
+    clientMetadataId,
+    clientToken,
+    components,
+    locale,
+    pageType,
+    partnerAttributionId,
+    shopperSessionId,
+    testBuyerCountry,
+    children,
+    scriptOptions,
+}) => {
+    // Auto-memoize props based on deep equality to prevent unnecessary reloads
+    const memoizedCreateOptions = useDeepCompareMemoize({
+        clientMetadataId,
+        clientToken,
+        components,
+        locale,
+        pageType,
+        partnerAttributionId,
+        shopperSessionId,
+        testBuyerCountry,
+    });
+    const memoizedScriptOptions = useDeepCompareMemoize(scriptOptions);
+
+    // Track if we've already handled the initial hydration
+    const hasHandledInitialHydration = useRef(false);
+
+    const [state, dispatch] = useReducer(instanceReducer, {
+        sdkInstance: null,
+        eligiblePaymentMethods: null,
+        loadingStatus: INSTANCE_LOADING_STATE.INITIAL,
+        error: null,
+        createInstanceOptions: memoizedCreateOptions,
+        scriptOptions: memoizedScriptOptions,
+    });
+
+    // Client-side hydration: transition from INITIAL to PENDING state
+    useEffect(() => {
+        if (
+            !isServer() &&
+            state.loadingStatus === INSTANCE_LOADING_STATE.INITIAL &&
+            !hasHandledInitialHydration.current
+        ) {
+            hasHandledInitialHydration.current = true;
+            dispatch({
+                type: INSTANCE_DISPATCH_ACTION.SET_LOADING_STATUS,
+                value: INSTANCE_LOADING_STATE.PENDING,
+            });
+        }
+    }, [state.loadingStatus]); // Run when loadingStatus changes, but only act once
+
+    // Auto-sync createInstanceOptions changes (e.g., client token updates)
+    useEffect(() => {
+        const hasOptionsChanged =
+            state.createInstanceOptions !== memoizedCreateOptions ||
+            state.scriptOptions !== memoizedScriptOptions;
+
+        if (hasOptionsChanged) {
+            dispatch({
+                type: INSTANCE_DISPATCH_ACTION.RESET_STATE,
+                value: {
+                    createInstanceOptions: memoizedCreateOptions,
+                    scriptOptions: memoizedScriptOptions,
+                },
+            });
+        }
+    }, [
+        memoizedCreateOptions,
+        memoizedScriptOptions,
+        state.createInstanceOptions,
+        state.scriptOptions,
```

Comment:
I'm a little confused by this logic. Why are we listening to both options for changes?